File: llm/trae/session.py
"""Trae LLM 会话管理.

管理 Trae LLM 客户端的会话上下文，支持持久化到 SQLite。
通过 flowforge MemoryManager 进行持久化，不直接操作数据库（铁律4）。
"""

from __future__ import annotations

import json
import time
from typing import Any, Dict, List, Optional

from flowforge.core.tracing import get_logger

from flowforge.llm.trae.config import TraeConfig

logger = get_logger("trae_llm.session")

# MemoryManager 的 short_term 存储使用的 key 前缀
_SESSION_KEY_PREFIX = "trae_session:"


class TraeSession:
    """Trae LLM 会话上下文.

    维护一个会话的消息历史，支持持久化到 SQLite（通过 MemoryManager）。
    每个会话有唯一的 session_id，用于关联 devforge 的任务。
    """

    def __init__(self, session_id: str, config: TraeConfig):
        self.session_id = session_id
        self._config = config
        self._messages: List[Dict[str, str]] = []
        self._created_at: float = time.time()
        self._updated_at: float = time.time()
        self._memory_manager: Optional[Any] = None

    def add_message(self, role: str, content: str) -> None:
        """添加消息到会话历史.

        Args:
            role: 消息角色 (system | user | assistant)
            content: 消息内容
        """
        if role not in ("system", "user", "assistant"):
            raise ValueError(f"role 必须是 system/user/assistant，得到: {role}")
        self._messages.append({"role": role, "content": content})
        self._updated_at = time.time()
        logger.debug(
            f"Session {self.session_id} 添加消息: role={role}, "
            f"len={len(content)}, total_messages={len(self._messages)}"
        )

    def get_context(self) -> List[Dict[str, str]]:
        """获取会话上下文（role/content 列表）.

        Returns:
            消息列表的副本。
        """
        return list(self._messages)

    def clear(self) -> None:
        """清除会话历史."""
        self._messages.clear()
        self._updated_at = time.time()
        logger.debug(f"Session {self.session_id} 已清除")

    def set_memory_manager(self, memory_manager: Any) -> None:
        """设置 MemoryManager 用于持久化（依赖注入，铁律3）.

        Args:
            memory_manager: flowforge MemoryManager 实例
        """
        self._memory_manager = memory_manager

    async def save(self) -> None:
        """持久化会话到 SQLite（通过 MemoryManager，不直接操作数据库）.

        如果未配置 session_persistence 或未注入 MemoryManager，则跳过。
        """
        if not self._config.session_persistence:
            return
        if self._memory_manager is None:
            logger.debug(
                f"Session {self.session_id} 未注入 MemoryManager，跳过持久化"
            )
            return
        try:
            data = {
                "session_id": self.session_id,
                "messages": self._messages,
                "created_at": self._created_at,
                "updated_at": self._updated_at,
            }
            key = f"{_SESSION_KEY_PREFIX}{self.session_id}"
            await self._memory_manager.save("short_term", key, data)
            logger.debug(f"Session {self.session_id} 已持久化 ({len(self._messages)} 条消息)")
        except Exception as e:
            logger.warning(f"Session {self.session_id} 持久化失败: {e}")

    async def load(self) -> None:
        """从持久化加载会话.

        如果未配置 session_persistence 或未注入 MemoryManager，则跳过。
        """
        if not self._config.session_persistence:
            return
        if self._memory_manager is None:
            logger.debug(
                f"Session {self.session_id} 未注入 MemoryManager，跳过加载"
            )
            return
        try:
            key = f"{_SESSION_KEY_PREFIX}{self.session_id}"
            results = await self._memory_manager.retrieve("short_term", key)
            if results:
                # retrieve 返回的是 search 结果列表，取第一个匹配的
                item = results[0]
                data = item.get("value", item) if isinstance(item, dict) else item
                if isinstance(data, dict) and "messages" in data:
                    self._messages = data["messages"]
                    self._created_at = data.get("created_at", self._created_at)
                    self._updated_at = data.get("updated_at", self._updated_at)
                    logger.debug(
                        f"Session {self.session_id} 已加载 ({len(self._messages)} 条消息)"
                    )
        except Exception as e:
            logger.warning(f"Session {self.session_id} 加载失败: {e}")

    def to_dict(self) -> Dict[str, Any]:
        """序列化会话为字典."""
        return {
            "session_id": self.session_id,
            "messages": self._messages,
            "created_at": self._created_at,
            "updated_at": self._updated_at,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any], config: TraeConfig) -> TraeSession:
        """从字典反序列化会话."""
        session = cls(data["session_id"], config)
        session._messages = data.get("messages", [])
        session._created_at = data.get("created_at", time.time())
        session._updated_at = data.get("updated_at", time.time())
        return session


class TraeSessionManager:
    """Trae LLM 会话管理器.

    管理多个 TraeSession 实例，支持创建、获取、关闭会话。
    """

    def __init__(self, config: Optional[TraeConfig] = None):
        self._config = config or TraeConfig()
        self._sessions: Dict[str, TraeSession] = {}
        self._memory_manager: Optional[Any] = None

    def set_memory_manager(self, memory_manager: Any) -> None:
        """注入 MemoryManager（依赖注入，铁律3）.

        设置后，所有新建的会话都会自动使用该 MemoryManager 进行持久化。
        """
        self._memory_manager = memory_manager

    def create_session(self, session_id: str) -> TraeSession:
        """创建新会话.

        如果 session_id 已存在，返回现有会话。

        Args:
            session_id: 会话唯一标识（如 devforge:coder:task123）

        Returns:
            TraeSession 实例
        """
        if session_id in self._sessions:
            logger.debug(f"Session {session_id} 已存在，返回现有会话")
            return self._sessions[session_id]
        session = TraeSession(session_id, self._config)
        if self._memory_manager is not None:
            session.set_memory_manager(self._memory_manager)
        self._sessions[session_id] = session
        logger.info(f"创建会话: {session_id}")
        return session

    def get_session(self, session_id: str) -> Optional[TraeSession]:
        """获取会话.

        Args:
            session_id: 会话唯一标识

        Returns:
            TraeSession 实例，不存在则返回 None
        """
        return self._sessions.get(session_id)

    def close_session(self, session_id: str) -> None:
        """关闭会话（从内存中移除）.

        注意：不会删除已持久化的数据。

        Args:
            session_id: 会话唯一标识
        """
        if session_id in self._sessions:
            del self._sessions[session_id]
            logger.info(f"关闭会话: {session_id}")

    def list_sessions(self) -> List[str]:
        """列出所有活跃会话 ID."""
        return list(self._sessions.keys())

    async def close_all(self) -> None:
        """关闭所有会话."""
        self._sessions.clear()
        logger.info("所有会话已关闭")

File: packages/forgekin/trae-bridge/src/session.ts
/**
 * @flowforge/forgekin-trae-bridge — 会话管理（TS 移植自 `llm/trae/session.py`）
 *
 * 管理 Trae LLM 客户端的会话上下文，支持通过注入的 SessionMemoryStore
 * 持久化（对齐 Python MemoryManager 依赖注入，铁律 3/4：不直接操作数据库）。
 */
import type { BridgeMessage } from './models.js';
import type { TraeClientConfig } from './config.js';

/** MemoryManager 的 short_term 存储使用的 key 前缀 */
const SESSION_KEY_PREFIX = 'trae_session:';

/** 会话持久化存储接口（对齐 Python MemoryManager save/retrieve 关键接口） */
export interface SessionMemoryStore {
  save(scope: string, key: string, value: unknown): Promise<void>;
  retrieve(scope: string, key: string): Promise<unknown[]>;
}

/** 会话序列化数据 */
export interface TraeSessionData {
  session_id: string;
  messages: BridgeMessage[];
  created_at: number;
  updated_at: number;
}

/** Trae LLM 会话上下文（维护消息历史 + 可选持久化） */
export class TraeSession {
  readonly sessionId: string;
  private readonly config: TraeClientConfig;
  private messages: BridgeMessage[] = [];
  private createdAt: number;
  private updatedAt: number;
  private memoryStore: SessionMemoryStore | null = null;

  constructor(sessionId: string, config: TraeClientConfig) {
    this.sessionId = sessionId;
    this.config = config;
    this.createdAt = Date.now();
    this.updatedAt = Date.now();
  }

  /** 添加消息到会话历史（role 非法时抛错，对齐 Python ValueError） */
  addMessage(role: string, content: string): void {
    if (role !== 'system' && role !== 'user' && role !== 'assistant') {
      throw new TypeError(`role 必须是 system/user/assistant，得到: ${role}`);
    }
    this.messages.push({ role, content });
    this.updatedAt = Date.now();
  }

  /** 获取会话上下文（消息列表的副本） */
  getContext(): BridgeMessage[] {
    return [...this.messages];
  }

  /** 清除会话历史 */
  clear(): void {
    this.messages = [];
    this.updatedAt = Date.now();
  }

  /** 设置持久化存储（依赖注入，铁律 3） */
  setMemoryStore(memoryStore: SessionMemoryStore): void {
    this.memoryStore = memoryStore;
  }

  /** 持久化会话（未配置 session_persistence 或未注入存储时跳过；失败仅吞掉） */
  async save(): Promise<void> {
    if (!this.config.session_persistence) {
      return;
    }
    if (this.memoryStore === null) {
      return;
    }
    try {
      const data: TraeSessionData = {
        session_id: this.sessionId,
        messages: this.messages,
        created_at: this.createdAt,
        updated_at: this.updatedAt,
      };
      await this.memoryStore.save('short_term', `${SESSION_KEY_PREFIX}${this.sessionId}`, data);
    } catch {
      // 对齐 Python：持久化失败仅 warning，不影响主流程
    }
  }

  /** 从持久化加载会话（未配置/未注入时跳过；失败仅吞掉） */
  async load(): Promise<void> {
    if (!this.config.session_persistence) {
      return;
    }
    if (this.memoryStore === null) {
      return;
    }
    try {
      const results = await this.memoryStore.retrieve(
        'short_term',
        `${SESSION_KEY_PREFIX}${this.sessionId}`,
      );
      if (results.length > 0) {
        // retrieve 返回的是搜索结果列表，取第一个匹配（对齐 Python results[0]）
        const item = results[0];
        let data: unknown = item;
        if (typeof item === 'object' && item !== null && 'value' in item) {
          data = (item as Record<string, unknown>)['value'];
        }
        if (typeof data === 'object' && data !== null && 'messages' in data) {
          const record = data as Record<string, unknown>;
          const messages = record['messages'];
          if (Array.isArray(messages)) {
            this.messages = messages as BridgeMessage[];
          }
          if (typeof record['created_at'] === 'number') {
            this.createdAt = record['created_at'];
          }
          if (typeof record['updated_at'] === 'number') {
            this.updatedAt = record['updated_at'];
          }
        }
      }
    } catch {
      // 对齐 Python：加载失败仅 warning
    }
  }

  /** 序列化会话为字典 */
  toDict(): TraeSessionData {
    return {
      session_id: this.sessionId,
      messages: this.messages,
      created_at: this.createdAt,
      updated_at: this.updatedAt,
    };
  }

  /** 从字典反序列化会话 */
  static fromDict(data: TraeSessionData, config: TraeClientConfig): TraeSession {
    const session = new TraeSession(data.session_id, config);
    session.messages = data.messages ?? [];
    session.createdAt = data.created_at ?? Date.now();
    session.updatedAt = data.updated_at ?? Date.now();
    return session;
  }
}

/** Trae LLM 会话管理器（管理多个 TraeSession 实例） */
export class TraeSessionManager {
  private readonly config: TraeClientConfig;
  private readonly sessions: Map<string, TraeSession> = new Map();
  private memoryStore: SessionMemoryStore | null = null;

  constructor(config: TraeClientConfig) {
    this.config = config;
  }

  /** 注入持久化存储（设置后所有新建会话自动使用） */
  setMemoryStore(memoryStore: SessionMemoryStore): void {
    this.memoryStore = memoryStore;
  }

  /** 创建新会话（已存在时返回现有会话，对齐 Python create_session） */
  createSession(sessionId: string): TraeSession {
    const existing = this.sessions.get(sessionId);
    if (existing !== undefined) {
      return existing;
    }
    const session = new TraeSession(sessionId, this.config);
    if (this.memoryStore !== null) {
      session.setMemoryStore(this.memoryStore);
    }
    this.sessions.set(sessionId, session);
    return session;
  }

  /** 获取会话（不存在返回 null） */
  getSession(sessionId: string): TraeSession | null {
    return this.sessions.get(sessionId) ?? null;
  }

  /** 关闭会话（从内存中移除，不删除已持久化的数据） */
  closeSession(sessionId: string): void {
    this.sessions.delete(sessionId);
  }

  /** 列出所有活跃会话 ID */
  listSessions(): string[] {
    return [...this.sessions.keys()];
  }

  /** 关闭所有会话 */
  async closeAll(): Promise<void> {
    this.sessions.clear();
  }
}
